Sources/Datadog/Tracing/Span/SpanTagsReducer.swift

```diff
@@ -0,0 +1,69 @@
+    static let errorMessage  = "error.msg"
+    static let errorStack    = "error.stack"
+}
+
+/// Reduces `DDSpan` tags and log attributes by extracting values that require separate handling.
+///
+/// The responsibility of `SpanTagsReducer` is to capture Open Tracing [tags](https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table)
+/// and [log fields](https://github.com/opentracing/specification/blob/master/semantic_conventions.md#log-fields-table) given by the user and
+/// transform them into the format used by Datadog. This happens in two ways, by:
+/// - extracting information from `spanTags` and `logFields` to `extracted*` variables,
+/// - reducing the initial `spanTags` collection by removing extracted information.
+///
+/// In result, the `spanTags` will contain only the tags that do NOT require special handling by Datadog and can be send as generic `span.meta.*` JSON values.
+/// Values extracted from `spanTags` and `logFields` will be passed to the `Span` encoding process in a type-safe manner.
+internal struct SpanTagsReducer {
+    /// Tags received by the user. This collection may be changed by the `reduce(_:)` function.
+    private(set) var spanTags: [String: Codable]
+    /// Log fieds send by the user for this `DDSpan`. This collection is never mutated, and is only used to extract extra information about the span.
+    private let logFields: [[String: Codable]]
+
+    // MARK: - Extracted Info
+
+    private(set) var extractedIsError: Bool? = nil
+
+    // MARK: - Reducers
+
+    internal static func reduce(spanTags: [String: Codable], logFields: [[String: Codable]]) -> SpanTagsReducer {
+        var reducer = SpanTagsReducer(spanTags: spanTags, logFields: logFields)
+        reducers.forEach { reduce in reduce(&reducer) }
+        return reducer
+    }
+
+    private static let reducers: [(inout SpanTagsReducer) -> Void] = [
+        extractErrorFromLogFields,
+        extractErrorFromTags
+    ]
+
+    private static func extractErrorFromLogFields(_ reducer: inout SpanTagsReducer) {
+        for fields in reducer.logFields {
+            let isErrorEvent = fields[OpenTracingLogFields.event] as? String == "error"
+            let errorKind = fields[OpenTracingLogFields.errorKind] as? String
+
+            if isErrorEvent || errorKind != nil {
+                reducer.extractedIsError = true
+                reducer.spanTags[DatadogTagKeys.errorMessage] = fields[OpenTracingLogFields.message] as? String
+                reducer.spanTags[DatadogTagKeys.errorType] = errorKind
+                reducer.spanTags[DatadogTagKeys.errorStack] = fields[OpenTracingLogFields.stack] as? String
+                return // ignore further logs
+            }
+        }
+    }
+
+    private static func extractErrorFromTags(_ reducer: inout SpanTagsReducer) {
+        if reducer.spanTags[OpenTracingTagKeys.error] as? Bool == true {
+            reducer.extractedIsError = true
+        }
+    }
+}
```

Comment:
this file is quite hard to understand IMO: 
• where do we mutate? 
• why do we reduce the reducer `reducers.forEach { reduce in reduce(&reducer) }`?
• we pass `spanTags` to reducer and then we use `reducer.spanTags`, it's not that clear what's changed?
• is the order of `reducers` important?

if i understood well, we take `logFields` and construct a special set of `spanTags` and we extract various info from `spanTags`
pseudo-code:
```swift
let logFields = someLogFields()
let spanTags = someSpanTags()

let datadogEncodedSpanTags = spanTags.merge(with: logsFields.datadogSpanTags)
let hasError = datadogEncodedSpanTags.hasError
let resourceName = datadogEncodedSpanTags.extractResourceName() // mutating func
```

above, `datadogSpanTags`, `hasError` and `extractResourceName()` can be implemented in extensions for `spanTag` and `logFields` types (different kinds of `Dict`s)

what do you think?

Reply:
Yes, this piece of logic is tricky. Generally speaking, this problem can be simplified to a **pure function**:
```swift
func computeStateBasedOn(dictionaryA: [String: Codable], dictionaryB: [String: Codable]) -> State
```
We take the `logFields` and `spanTags` (dictionaries `A` and `B`)  and compute a `State` which describes following:
* `spanTags` (may be different than the input `spanTags`),
* `isError` flag,
* `resourceName` value,
* other values in the future (likely: HTTP status codes).

And what we do now, conforms to a pure function signature:
```swift
let state = SpanTagsReducer.reduce(spanTags: ddspan.tags, logFields: ddspan.logFields)
_ = state.spanTags
_ = state.extractedIsError
_ = state.extractedResourceName
```

We can't merge the dictionaries initially, as proposed:
```swift
let datadogEncodedSpanTags = spanTags.merge(with: logsFields.datadogSpanTags)
```
because it is important from which dictionary, the entry comes from (1) and both dictionaries may contain the same keys (2). In both cases, we will lost the information.

Also, we can't provide mutating getters, as proposed:
```swift
let resourceName = datadogEncodedSpanTags.extractResourceName() // mutating func
```
because state computation changes the input dictionaries, so this will give a different results depending on the order of method calling:
```swift
var mergedDictionaries = spanTags.merge(with: logsFields.datadogSpanTags)

// this:
let datadogEncodedSpanTags = mergedDictionaries
let hasError = mergedDictionaries.hasError

// will not mean the same as this:
let hasError = mergedDictionaries.hasError
let datadogEncodedSpanTags = mergedDictionaries
```

And lastly, IMO the `.isError` extension to a `[String: Codable]` type is a very bad practise - it will add the `.isError` interface on all other dictionaries of this kind (even if it's declared privately in the same file, it's kind of abuse IMO).

Reply:
the usage of this type is 👍 
however, its implementation seems to be over-abstracted for now
we can start simple, without `reducers` array and `inout` parameters
then we can progress when needed

Reply:
Fixed in https://github.com/DataDog/dd-sdk-ios/pull/121/commits/b49d1b1b98ea135bf26ec4e9b6ae447dfe3b4a02, #121 to avoid rebasing 🙌.